### ISSUE TITLE
PHP Fatal error:  Uncaught InvalidArgumentException: Matrix must contain at least 1 element.

### ISSUE THREAD

**Issue by ghost** [RubixML/ML#92]
**Describe the bug**
PHP Fatal error when training my first estimator with Rubix ML.

**To Reproduce**
Run the [MLP Regressor](https://docs.rubixml.com/en/latest/regressors/mlp-regressor.html) example:
```php
// cli/train.php

require_once __DIR__ . '/../vendor/autoload.php';

use Rubix\ML\Datasets\Labeled;
use Rubix\ML\Regressors\MLPRegressor;
use Rubix\ML\NeuralNet\CostFunctions\LeastSquares;
use Rubix\ML\NeuralNet\Layers\Dense;
use Rubix\ML\NeuralNet\Layers\Activation;
use Rubix\ML\NeuralNet\ActivationFunctions\ReLU;
use Rubix\ML\NeuralNet\Optimizers\RMSProp;
use Rubix\ML\CrossValidation\Metrics\RSquared;

$estimator = new MLPRegressor([
    new Dense(100),
    new Activation(new ReLU()),
    new Dense(100),
    new Activation(new ReLU()),
    new Dense(50),
    new Activation(new ReLU()),
    new Dense(50),
    new Activation(new ReLU()),
], 128, new RMSProp(0.001), 1e-3, 100, 1e-5, 3, 0.1, new LeastSquares(), new RSquared());


$samples = [
    [0.1, 20, 1],
    [2.0, -5, 2],
    [0.01, 5, 1],
];

$labels = [0.5, 0.6, 0.7];

$dataset = new Labeled($samples, $labels);

$estimator->train($dataset);
```

**Expected behavior**
The estimator is successfully trained.

**Additional context**
```
$ docker exec -it pgn_chess_data_php_fpm php cli/train.php
PHP Fatal error:  Uncaught InvalidArgumentException: Matrix must contain at least 1 element. in /usr/share/nginx/pgn-chess-data/vendor/rubix/tensor/src/Matrix.php:444
Stack trace:
#0 /usr/share/nginx/pgn-chess-data/vendor/rubix/tensor/src/Matrix.php(70): Tensor\Matrix->__construct(Array, false)
#1 /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/NeuralNet/FeedForward.php(163): Tensor\Matrix::quick(Array)
#2 /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php(470): Rubix\ML\NeuralNet\FeedForward->infer(Object(Rubix\ML\Datasets\Labeled))
#3 /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php(403): Rubix\ML\Regressors\MLPRegressor->predict(Object(Rubix\ML\Datasets\Labeled))
#4 /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php(346): Rubix\ML\Regressors\MLPRegressor->partial(Object(Rubix\ML\Datasets\Labeled))
#5 /usr/share/nginx/pgn-chess-data/cli/train.php(36): Rubix\ML\Regressors\MLPRegressor->train(Object(Rubix\ML\Datasets\Labeled))
#6 {main in /usr/share/nginx/pgn-chess-data/vendor/rubix/tensor/src/Matrix.php on line 444

Fatal error: Uncaught InvalidArgumentException: Matrix must contain at least 1 element. in /usr/share/nginx/pgn-chess-data/vendor/rubix/tensor/src/Matrix.php:444
Stack trace:
#0 /usr/share/nginx/pgn-chess-data/vendor/rubix/tensor/src/Matrix.php(70): Tensor\Matrix->__construct(Array, false)
#1 /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/NeuralNet/FeedForward.php(163): Tensor\Matrix::quick(Array)
#2 /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php(470): Rubix\ML\NeuralNet\FeedForward->infer(Object(Rubix\ML\Datasets\Labeled))
#3 /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php(403): Rubix\ML\Regressors\MLPRegressor->predict(Object(Rubix\ML\Datasets\Labeled))
#4 /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php(346): Rubix\ML\Regressors\MLPRegressor->partial(Object(Rubix\ML\Datasets\Labeled))
#5 /usr/share/nginx/pgn-chess-data/cli/train.php(36): Rubix\ML\Regressors\MLPRegressor->train(Object(Rubix\ML\Datasets\Labeled))
#6 {main in /usr/share/nginx/pgn-chess-data/vendor/rubix/tensor/src/Matrix.php on line 444
```

**Comment by andrewdalpino** [comment]
Hey @programarivm I think the reason you're getting this error if because you have such a small training set (3 samples) ... it's trying to create a hold out set with 10% of the samples but 10% of 3 rounds to 0. Can you try training with more samples (> 10) and see if that works? If so, we can add a check to make sure the learner has enough data to proceed before training.

**Comment by andrewdalpino** [comment]
This should be fixed now in  https://github.com/RubixML/RubixML/commit/a2362448aee4e2ae0f7e676e3d1271012909ad02 ... you can install the latest master by requiring `dev-master` in your composer.json if you would like to test it out.

Thank you for the great bug report! Join our Telegram channel if you can https://t.me/RubixML

**Comment by ghost** [comment]
Thanks for the prompt response. The error is gone if using a sample with 10 elements. I suppose I was expecting a basic small sample would work as per the documentation.

```php
$samples = [
    [3, 4, 50.5],
    [1, 5, 24.7],
    [4, 4, 62.0],
    [3, 2, 31.1],
];

$labels = ['married', 'divorced', 'married', 'divorced'];
```

Now with the fix the `RuntimeException` message looks like a little bit more descriptive.

```
$ docker exec -it pgn_chess_data_php_fpm php cli/train.php
PHP Fatal error:  Uncaught RuntimeException: Dataset does not contain enough records to create a validation set with a hold out ratio of 0.1. in /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php:381
Stack trace:
#0 /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php(346): Rubix\ML\Regressors\MLPRegressor->partial(Object(Rubix\ML\Datasets\Labeled))
#1 /usr/share/nginx/pgn-chess-data/cli/train.php(36): Rubix\ML\Regressors\MLPRegressor->train(Object(Rubix\ML\Datasets\Labeled))
#2 {main}
  thrown in /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php on line 381

Fatal error: Uncaught RuntimeException: Dataset does not contain enough records to create a validation set with a hold out ratio of 0.1. in /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php:381
Stack trace:
#0 /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php(346): Rubix\ML\Regressors\MLPRegressor->partial(Object(Rubix\ML\Datasets\Labeled))
#1 /usr/share/nginx/pgn-chess-data/cli/train.php(36): Rubix\ML\Regressors\MLPRegressor->train(Object(Rubix\ML\Datasets\Labeled))
#2 {main}
  thrown in /usr/share/nginx/pgn-chess-data/vendor/rubix/ml/src/Regressors/MLPRegressor.php on line 381
```

**Comment by andrewdalpino** [comment]
Great, I'll add a note to the documentation about this and consider it resolved!

We use the validation set for progress monitoring, early stopping, and network snapshotting in case you were wondering

Thanks again for the bug report!

**Comment by andrewdalpino** [comment]
P.S. In the future, we'll make it so that if it won't use a validation set (no early stopping or snapshotting) if there isn't enough data. This should allow internal CV learners to train on datasets as small as a single record.

See https://github.com/RubixML/RubixML/issues/95
